From bfec0f11eadd7ed3597bc9b550a5595421a4cb31 Mon Sep 17 00:00:00 2001
From: Alan Modra <amodra@gmail.com>
Date: Sun, 1 Oct 2017 21:40:23 +1030
Subject: [PATCH] PR22232, NULL pointer dereference in
 load_specific_debug_section

	PR 22232
	PR 22230
	* objdump.c (load_specific_debug_section): Introduce a temp to
	stop bfd_get_full_section_contents NULLing out section->start.
---
 binutils/ChangeLog |    7 +++++++
 binutils/objdump.c |    5 +++--
 2 files changed, 10 insertions(+), 2 deletions(-)

diff --git a/binutils/ChangeLog b/binutils/ChangeLog
index d45eeaf..95cf50b 100644
--- a/binutils/ChangeLog
+++ b/binutils/ChangeLog
@@ -50,6 +50,13 @@
 
 2017-10-01  Alan Modra  <amodra@gmail.com>
 
+	PR 22232
+	PR 22230
+	* objdump.c (load_specific_debug_section): Introduce a temp to
+	stop bfd_get_full_section_contents NULLing out section->start.
+
+2017-10-01  Alan Modra  <amodra@gmail.com>
+
 	PR 22230
 	* objdump.c (load_specific_debug_section): Allocate an extra byte
 	for a terminating NUL.
diff --git a/binutils/objdump.c b/binutils/objdump.c
index 83b8b2a..98c316a 100644
--- a/binutils/objdump.c
+++ b/binutils/objdump.c
@@ -2313,6 +2313,7 @@ load_specific_debug_section (enum dwarf_
 {
   struct dwarf_section *section = &debug_displays [debug].section;
   bfd *abfd = (bfd *) file;
+  bfd_byte *contents;
   bfd_boolean ret;
 
   /* If it is already loaded, do nothing.  */
@@ -2323,10 +2324,10 @@ load_specific_debug_section (enum dwarf_
   section->num_relocs = 0;
   section->address = bfd_get_section_vma (abfd, sec);
   section->size = bfd_get_section_size (sec);
-  section->start = malloc (section->size + 1);
+  section->start = contents = malloc (section->size + 1);
   section->user_data = sec;
   if (section->start == NULL
-      || !bfd_get_full_section_contents (abfd, sec, &section->start))
+      || !bfd_get_full_section_contents (abfd, sec, &contents))
     {
       free_debug_section (debug);
       printf (_("\nCan't get contents for section '%s'.\n"),
-- 
1.7.9.5

